src/main/java/io/confluent/connect/hdfs/wal/FSWAL.java

```diff
@@ -183,13 +189,160 @@ private void commitEntriesToStorage(Map<WALEntry, WALEntry> entries) {
     }
   }
 
+  /**
+   * Extract the latest offset and file path from the WAL file.
+   * Attempt with the most recent WAL file, and fall back to the old file if it's not applicable.
+   *
+   * <p> The old WAL is used when the most recent WAL file has already been truncated
+   * and is not existent, which may happen when the connector has flushed all records,
+   * and is restarted. </p>
+   *
+   * @return the latest offset and filepath from the WAL file or null
+   */
+  public Pair<Long, String> extractLatestOffsetFromWAL() {
+    String oldWALFile = logFile + OLD_LOG_EXTENSION;
+    try {
+      Pair<Long, String> latestOffset = null;
+      if (storage.exists(logFile)) {
+        log.trace("Restoring offset from WAL file: {}", logFile);
+        if (reader == null) {
+          reader = newWalFileReader();
+        } else {
+          // reset read position after apply()
+          reader.seekToFirstRecord();
+        }
+        List<String> committedFileBatch = getLastFilledBlockFromWAL(reader);
+        // At this point the committedFilenames list will contain the
+        // filenames in the last BEGIN-END block of the file. Find the latest offsets among these.
+        latestOffset = getLatestOffsetFromList(committedFileBatch);
+      }
+
+      // attempt to use old log file if recent WAL is empty or non-existent
+      if (latestOffset == null && storage.exists(oldWALFile)) {
+        log.trace("Could not find offset in log file {}. Using {} instead", logFile, oldWALFile);
+        try (Reader oldFileReader =
+            new WALFile.Reader(conf.getHadoopConfiguration(), Reader.file(new Path(oldWALFile)))
+        ) {
+          List<String> committedFileBatch = getLastFilledBlockFromWAL(oldFileReader);
+          latestOffset = getLatestOffsetFromList(committedFileBatch);
+        }
+      }
+      return latestOffset;
+    } catch (IOException e) {
+      log.warn(
+          "Error restoring offsets from either {} or {} WAL files: {}",
+          logFile,
+          oldWALFile,
+          e.getMessage()
+      );
+      return null;
+    }
+  }
+
+  /**
+   * Extract the last filled BEGIN-END block of entries from the WAL.
+   *
+   * @param reader the WAL file reader
+   * @return the last batch of entries, may be empty if the WAL
+   *         is empty or only contains empty BEGIN-END blocks
+   * @throws IOException error on reading the WAL file
+   */
+  private List<String> getLastFilledBlockFromWAL(Reader reader) throws IOException {
+    // In a WAL entry the temp filenames (keys) don't contain offset info,
+    // so we only need to track the committed filename (values).
+    ArrayList<String> committedFilenames = new ArrayList<>();
```

Comment:
@dosvath in some cases a comment applies to several instances in the file. I think I forgot to mention to replace the rest of the impl declarations with the interface. Please check for `ArrayList` declarations and use `List`

Reply:
Got it :)